README.md
# Howto

## Minikube

Start minikube with mutation admission webhook support.

```
minikube start
```

Old minikube/kubernetes < 1.9 will need this --extra-config:

```
minikube start --extra-config=apiserver.admission-control="NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota"
```

For inspecting the running system inside of minikube with docker, run:

```
eval $(minikube docker-env)
```

## Setup Kong

### Run Kong on k8s

```
git clone https://github.com/Kong/kong-dist-kubernetes.git
cd kong-dist-kubernetes
make run_<postgres|cassandra>
```

or

```
kubectl apply -f https://raw.githubusercontent.com/Kong/kong-dist-kubernetes/master/kong-namespace.yaml
curl -fsSL https://raw.githubusercontent.com/Kong/kong-dist-kubernetes/master/setup_certificate.sh | bash
kubectl -n kong apply -f https://raw.githubusercontent.com/Kong/kong-dist-kubernetes/master/postgres.yaml
kubectl -n kong apply -f https://raw.githubusercontent.com/Kong/kong-dist-kubernetes/master/kong-control-plane-postgres.yaml
kubectl -n kong apply -f https://raw.githubusercontent.com/Kong/kong-dist-kubernetes/master/kong-ingress-data-plane-postgres.yaml
```

Wait for the Kong control and data planes to be running

```
kubectl -n kong get all
```

### Turn on kong plugins

```
export HOST=$(kubectl get nodes --namespace default -o jsonpath='{.items[0].status.addresses[0].address}')
export ADMIN_PORT=$(kubectl get svc --namespace kong kong-control-plane  -o jsonpath='{.spec.ports[0].nodePort}')
curl $HOST:$ADMIN_PORT/plugins -d name=kubernetes-sidecar-injector
```

### Turn on sidecar injection

```
cat <<EOF | kubectl create -f -
apiVersion: admissionregistration.k8s.io/v1beta1
kind: MutatingWebhookConfiguration
metadata:
  name: kong-sidecar-injector
webhooks:
- name: kong.sidecar.injector
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations: [ "CREATE" ]
  failurePolicy: Fail
  namespaceSelector:
    matchExpressions:
    - key: kong-sidecar-injection
      operator: NotIn
      values:
      - disabled
  clientConfig:
    service:
      namespace: kong
      name: kong-control-plane
      path: /kubernetes-sidecar-injector
    caBundle: $(kubectl config view --raw --minify --flatten -o jsonpath='{.clusters[].cluster.certificate-authority-data}')
EOF
```

or

```
curl -fsSL https://raw.githubusercontent.com/Kong/kong-dist-kubernetes/master/setup_sidecar_injector.sh | bash
```

Going forward any pod's that start get a Kong sidecar automatically injected 
which all data from the containers of that pod will flow through

For example starting the bookinfo.yaml example

```
kubectl apply -f https://raw.githubusercontent.com/istio/istio/release-1.1/samples/bookinfo/platform/kube/bookinfo.yaml
```

Every pod has an additional Kong docker dataplane image running

```
NAME                                  READY   STATUS    RESTARTS   AGE
pod/details-v1-5df567f4f6-vfj5z       2/2     Running   0          2m53
```

## Debugging

### Logs from api server

These logs can be useful if the webhook doesn't seem to be getting called.

```
kubectl logs -n kube-system pod/kube-apiserver-minikube
```


### Logs from controller manager

Watch the logs of `kube-controller-manager-minikube` for JSONPatch issues

```
kubectl logs -n kube-system pod/kube-controller-manager-minikube
```


### Logs from kong

```
while true; do kubectl logs -n kong -f $(kubectl get pods -l app=kong-control-plane -n kong -o name); done
```


### Working on the plugin

```
docker build . -t kong && kubectl -n kong delete $(kubectl get pods -l app=kong-control-plane -n kong -o name)
```


NEWS
0.2.1 - 2019-06-25

  - Fix api to call process_auto_fields on input data before validating,
    that makes the validation more resilient to Kubernetes changes
    (see https://github.com/Kong/kubernetes-sidecar-injector/issues/7)
  - Fix api with nicer error reporting on object schema errors

  - Fix schema by adding `kind` and `apiVersion` to Pod schema
    (see https://github.com/Kong/kubernetes-sidecar-injector/issues/7)

0.2.0 - 2019-06-07

  - Remove `BasePlugin` dependency (not needed anymore)


0.1.1 - 2019-03-11

  - Add `name` to the plugin handler to make Kong logs more readable


0.1.0 - 2019-02-13

  - Initial release


kong/plugins/kubernetes-sidecar-injector/config.lua
-- Loads a plugin config from the datastore.
-- @return plugin config table or an empty sentinel table in case of a db-miss
local function load_plugin_into_memory(key)
  local row, err = kong.db.plugins:select_by_cache_key(key)
  if err then
    return nil, tostring(err)
  end

  return row
end


--- Get the configuration for a plugin entry in the DB.
-- Given a Route, Service, Consumer and a plugin name, retrieve the plugin's
-- configuration if it exists. Results are cached in ngx.dict
-- @param[type=string] route_id ID of the route being proxied.
-- @param[type=string] service_id ID of the service being proxied.
-- @param[type=string] consumer_id ID of the Consumer making the request (if any).
-- @param[type=stirng] plugin_name Name of the plugin being tested for.
-- @treturn table Plugin retrieved from the cache or database.
local function get_plugin_configuration(plugin_name,
                                        route_id,
                                        service_id,
                                        consumer_id)
  local key = kong.db.plugins:cache_key(plugin_name,
                                        route_id,
                                        service_id,
                                        consumer_id)

  local plugin, err = kong.cache:get(key, nil, load_plugin_into_memory, key)
  if err then
    ngx.log(ngx.ERR, tostring(err))
    return ngx.exit(ngx.ERROR)
  end

  if not plugin or not plugin.enabled then
    return nil
  end

  return plugin.config or {}
end

return {
  get_plugin_configuration = get_plugin_configuration,
}


kong/plugins/kubernetes-sidecar-injector/schema.lua
local typedefs = require "kong.db.schema.typedefs"
local k8s_typedefs = require "kong.plugins.kubernetes-sidecar-injector.typedefs"

return {
  name = "kubernetes-sidecar-injector",
  fields = {
    -- This plugin should only be loaded globally
    { route = typedefs.no_route },
    { service = typedefs.no_service },
    { consumer = typedefs.no_consumer },

    { config = { type = "record", fields = {
        { initImage = { type = "string", default = "istio/proxy_init:1.0.5" } },
        { initImagePullPolicy = k8s_typedefs.ImagePullPolicy { default = "IfNotPresent" } },
        { initArgs = { type = "array", elements = { type = "string" }, default = {
          "-p", "7000",
          "-u", "1337",
          "-m", "TPROXY",
          "-i", "*",
          "-b", "*",
        } } },
        { image = { type = "string", default = "kong" } },
        { imagePullPolicy = k8s_typedefs.ImagePullPolicy { default = "IfNotPresent" } },
        { extra_env = { type = "map",
          keys = { type = "string" },
          values = { type = "string" }
        } },
        { http_port = typedefs.port { default = 8000 } },
        { https_port = typedefs.port { default = 8443 } },
        { stream_port = typedefs.port { default = 7000 } }, -- should match initArgs default
        { namespace_blacklist = { type = "array", elements = { type = "string" },
                                  default = { "kube-system" } } },
    } } },
  },
}


kong/plugins/kubernetes-sidecar-injector/typedefs.lua
local Schema = require "kong.db.schema"
local typedefs = require "kong.db.schema.typedefs"


local int32 = Schema.define { type = "integer", between = { -2147483648, 2147483647 } }

-- XXX: 9223372036854775807 gets rounded by lua < 5.3. Can we use LuaJIT long longs?
local int64 = Schema.define { type = "integer", between = { -9223372036854775808, 9223372036854775807 } }

-- represented in RFC3339 form and is in UTC
local Time = Schema.define { type = "string" }


--- Kubernetes Object definitions
-- These can be figured out via reading the kubernetes source.

-- https://github.com/kubernetes/kubernetes/blob/v1.13.1/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

local OwnerReference = Schema.define { type = "record", fields = {
  { apiVersion = { type = "string", required = true } },
  { kind = { type = "string", required = true } },
  { name = { type = "string", required = true } },
  { uid = typedefs.uuid { required = true } },
  { controller = { type = "boolean" } },
  { blockOwnerDeletion = { type = "boolean" } },
} }

local ListMeta = Schema.define { type = "record", fields = {
  { selfLink = { type = "string" } },
  { resourceVersion = { type = "string" } },
  { continue = { type = "string" } },
} }

local StatusReason = Schema.define { type = "string", len_min = 0, one_of = {
  "",
  "Unauthorized",
  "Forbidden",
  "NotFound",
  "AlreadyExists",
  "Conflict",
  "Gone",
  "Invalid",
  "ServerTimeout",
  "Timeout",
  "TooManyRequests",
  "BadRequest",
  "MethodNotAllowed",
  "NotAcceptable",
  "UnsupportedMediaType",
  "InternalError",
  "Expired",
  "ServiceUnavailable",
} }

local CauseType = Schema.define { type = "string", one_of = {
  "FieldValueNotFound",
  "FieldValueRequired",
  "FieldValueDuplicate",
  "FieldValueInvalid",
  "FieldValueNotSupported",
  "UnexpectedServerResponse",
} }

local StatusCause = Schema.define { type = "record", fields = {
  { reason = CauseType },
  { message = { type = "string" } },
  { field = { type = "string" } },
} }

local StatusDetails = Schema.define { type = "record", fields = {
  { name = { type = "string" } },
  { group = { type = "string" } },
  { kind = { type = "string" } },
  { uid = typedefs.uuid { } },
  { causes = { type = "array", elements = StatusCause } },
  { retryAfterSeconds = int32 },
} }

local Status = Schema.define { type = "record", fields = {
  { metadata = { type = ListMeta } },
  { status = { type = "string"  } },
  { message = { type = "string"  } },
  { reason = { type = StatusReason } },
  { details = { type = StatusDetails } },
  { code = int32 },
} }

local Initializer = Schema.define { type = "record", fields = {
  { name = { type = "string", required = true } },
} }

local Initializers = Schema.define { type = "record", fields = {
  { pending = { type = "array", required = true, elements = Initializer } },
  { status = Status { required = true } },
} }

local ObjectMeta = Schema.define { type = "record", fields = {
  { name = { type = "string" } },
  { generateName = { type = "string" } },
  { namespace = { type = "string" } },
  { selfLink = { type = "string" } },
  { uid = typedefs.uuid },
  { resourceVersion = { type = "string" } },
  { generation = int64 },
  { creationTimestamp = Time },
  { deletionTimestamp = Time },
  { deletionGracePeriodSeconds = { type = "integer" } },
  { labels = { type = "map", keys = { type = "string" }, values = { type = "string" } } },
  { annotations = { type = "map", keys = { type = "string" }, values = { type = "string" } } },
  { ownerReferences = { type = "array", elements = OwnerReference } },
  { initializers = Initializers },
  { finalizers = { type = "array", elements = { type = "string" } } },
  { clusterName = { type = "string" } },
} }


-- https://github.com/kubernetes/kubernetes/blob/v1.13.1/staging/src/k8s.io/apimachinery/pkg/runtime/types.go

local RawExtension = Schema.define { type = "any" }


-- https://github.com/kubernetes/kubernetes/blob/v1.13.1/staging/src/k8s.io/api/core/v1/types.go

local ImagePullPolicy = Schema.define { type = "string", one_of = {
  "Always",
  "Never",
  "IfNotPresent",
} }

-- TODO: Complete PodSpec definition
local PodSpec = Schema.define { type = "any" }

local Pod = Schema.define { type = "record", fields = {
  { metadata = ObjectMeta },
  { spec = PodSpec },
  { status = Status },
} }


-- https://github.com/kubernetes/kubernetes/blob/v1.13.1/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version.go

local GroupVersionKind = Schema.define { type = "record", fields = {
  { group = { type = "string", required = true, len_min = 0 } },
  { version = { type = "string", required = true } },
  { kind = { type = "string", required = true } },
} }

local GroupVersionResource = Schema.define { type = "record", fields = {
  { group = { type = "string", required = true, len_min = 0 } },
  { version = { type = "string", required = true } },
  { resource = { type = "string", required = true } },
} }


-- https://github.com/kubernetes/kubernetes/blob/v1.13.1/staging/src/k8s.io/api/authentication/v1/types.go

local UserInfo = Schema.define { type = "record", fields = {
  { username = { type = "string" } },
  { uid = typedefs.uuid },
  { groups = { type = "array", elements = { type = "string" } } },
  -- TODO: kong map type doesn't allow 'null' as value
  -- { extra = { type = "map", keys = { type = "string" }, values = { type = "any" } } },
  { extra = { type = "any" } },
} }


-- https://github.com/kubernetes/kubernetes/blob/v1.13.1/staging/src/k8s.io/api/admission/v1beta1/types.go

local Operation = Schema.define { type = "string", one_of = {
  "CREATE",
  "UPDATE",
  "DELETE",
  "CONNECT",
} }

local AdmissionRequest = Schema.define {
  type = "record",
  fields = {
    { uid = typedefs.uuid { required = true } },
    { kind = GroupVersionKind { required = true } },
    { resource = GroupVersionResource { required = true } },
    { subResource = { type = "string" } },
    { name = { type = "string" } },
    { namespace = { type = "string" } },
    { operation = Operation { required = true } },
    { userInfo = UserInfo { required = true } },
    { object = RawExtension },
    { oldObject = RawExtension },
    { dryRun = { type = "boolean" } },
  },
  entity_checks = {
    { conditional = { if_field = "operation",
      if_match = { type = "string", one_of = { "UPDATE" } },
      then_field = "oldObject",
      then_match = { required = true },
      then_err = "oldObject required for UPDATEs",
    } },
  },
}


return {
  OwnerReference = OwnerReference,
  ListMeta = ListMeta,
  StatusReason = StatusReason,
  CauseType = CauseType,
  StatusCause = StatusCause,
  StatusDetails = StatusDetails,
  Status = Status,
  Initializer = Initializer,
  Initializers = Initializers,
  ObjectMeta = ObjectMeta,
  RawExtension = RawExtension,
  ImagePullPolicy = ImagePullPolicy,
  PodSpec = PodSpec,
  Pod = Pod,
  GroupVersionKind = GroupVersionKind,
  GroupVersionResource = GroupVersionResource,
  UserInfo = UserInfo,
  Operation = Operation,
  AdmissionRequest = AdmissionRequest,
}


kong/plugins/kubernetes-sidecar-injector/api.lua
local cjson = require "cjson"
local Schema = require "kong.db.schema"
local kong_pdk = require "kong.pdk".new({}, 1)
local k8s_typedefs = require "kong.plugins.kubernetes-sidecar-injector.typedefs"
local load_configuration = require "kong.plugins.kubernetes-sidecar-injector.config"


local kong = kong
local fmt = string.format
local type = type
local pairs = pairs
local tconcat = table.concat
local tinsert = table.insert
local tostring = tostring
local log_info = kong_pdk.log.info
local encode_base64 = ngx.encode_base64


local function skip_injection(plugin_config, review_request) -- luacheck: ignore 212
  if type(review_request)                 == "table" and
     type(review_request.object)          == "table" and
     type(review_request.object.metadata) == "table" then

    local annotations = review_request.object.metadata.annotations

    if type(annotations) == "table" then
      -- Behave similar to sidecar.istio.io/inject annotation
      if annotations["k8s.konghq.com/sidecar-inject"] == "false" then
        return true
      end
    end
  end

  -- TODO: allow injection to be more configurable

  return false
end


-- Refine schema to what we actually want to accept
local podschema = Schema.new(k8s_typedefs.Pod)

local admissionreviewschema = Schema.new {
  name = "admissionregistration.k8s.io/v1beta1 AdmissionReview",
  fields = {
    { kind = { type = "string", eq = "AdmissionReview", required = true } },
    { apiVersion = { type = "string", eq = "admission.k8s.io/v1beta1", required = true } },
    { request = k8s_typedefs.AdmissionRequest { required = true, custom_validator = function(review_request)
      if review_request.operation ~= "CREATE" then
        return nil, "unsupported operation (this controller only understands the CREATE operation)"
      end

      local group_kind = review_request.kind
      if group_kind.group ~= "" or group_kind.version ~= "v1" or group_kind.kind ~= "Pod" then
        return nil, "unknown resource type (this controller only accepts v1 Pods)"
      end

      local object = podschema:process_auto_fields(review_request.object, "select", false)
      local ok, err = podschema:validate(object)
      if not ok then
        local err_t = kong.db.errors:schema_violation({ object = err })
        return nil, tostring(err_t)
      end

      if type(object.spec) ~= "table" then
        -- workaround for incomplete podspec definition
        return nil, "invalid object.spec field"
      end

      return true
    end } },
  }
}

return {
  ["/kubernetes-sidecar-injector"] = {
    schema = admissionreviewschema,
    methods = {
      POST = function(self)
        local plugin_config = load_configuration("kubernetes-sidecar-injector")
        -- 404 if plugin not found/enabled
        if not plugin_config then
          return kong.response.exit(404, { message = "Not found" })
        end

        -- TODO: only accept JSON?
        local args = admissionreviewschema:process_auto_fields(self.args.post, "select", false)
        local ok, err = admissionreviewschema:validate(args)
        if not ok then
          return kong.response.exit(422, { message = err })
        end

        local review_request = args.request
        local object = review_request.object

        -- Same log format as istio
        log_info("AdmissionReview for",
          " Kind=", review_request.kind.kind, "/", review_request.kind.version,
          " Namespace=", review_request.namespace,
          " Name=", review_request.name or "", " (", object.metadata.name or "", ")",
          " UID=", review_request.uid,
          " Rfc6902PatchOperation=", review_request.operation,
          -- XXX: even though required=true is set on the userInfo field, it can still be nil
          " UserInfo=", (review_request.userInfo or {}).username or ""
        )

        local reply = {
          kind = "AdmissionReview",
          apiVersion = "admission.k8s.io/v1beta1",
          response = { -- https://github.com/kubernetes/kubernetes/blob/v1.11.0/staging/src/k8s.io/api/admission/v1beta1/types.go#L77
            uid = review_request.uid,
            allowed = true,
            status = nil, -- only required if 'allowed' is false
            patch = nil,
            patchType = nil,
          }
        }

        if skip_injection(plugin_config, review_request) then
          log_info("Skipping ", review_request.namespace, "/", object.metadata.name, " due to policy check")
          return { json = reply }
        end

        -- Patches in RFC 6902 format
        local patches = { nil, nil }

        -- iptables setup container
        patches[1] = {
          op = "add",
          path = "/spec/initContainers/-",
          value = {
            name = "kong-iptables-setup",
            image = plugin_config.initImage,
            imagePullPolicy = plugin_config.initImagePullPolicy,
            args = plugin_config.initArgs,
            securityContext = { capabilities = { add = { "NET_ADMIN" } } },
          },
        }
        if not object.spec.initContainers then
          -- need to add array member instead of adding *to* it.
          patches[1].path = "/spec/initContainers"
          patches[1].value = { patches[1].value }
        end

        -- Add proxy sidecar container
        local config = kong.configuration
        local env = {
          -- disable admin interface in data plane
          { name = "KONG_ADMIN_LISTEN", value = "off" },
          { name = "KONG_PROXY_LISTEN", value =
            fmt("0.0.0.0:%d transparent", plugin_config.http_port) .. "," ..
            fmt("0.0.0.0:%d ssl transparent", plugin_config.https_port)
          },
          { name = "KONG_STREAM_LISTEN", value =
            fmt("0.0.0.0:%d transparent", plugin_config.stream_port)
          },
          { name = "KONG_PROXY_ACCESS_LOG", value = "/dev/stdout" },
          { name = "KONG_PROXY_ERROR_LOG", value = "/dev/stderr" },
          -- need to copy relevant database configuration
          { name = "KONG_DATABASE", value = config.database },
        }
        if config.database == "postgres" then
          tinsert(env, { name = "KONG_PG_HOST",
                         value = config.pg_host })
          tinsert(env, { name = "KONG_PG_PORT",
                         value = fmt("%d", config.pg_port) })
          tinsert(env, { name = "KONG_PG_TIMEOUT",
                         value = fmt("%d", config.pg_timeout) })
          tinsert(env, { name = "KONG_PG_USER",
                         value = config.pg_user })
          if config.pg_password then
            tinsert(env, { name = "KONG_PG_PASSWORD",
                           value = config.pg_password })
          end
          tinsert(env, { name = "KONG_PG_DATABASE",
                         value = config.pg_database })
          if config.pg_schema then
            tinsert(env, { name = "KONG_PG_SCHEMA",
                           value = config.pg_schema })
          end
          tinsert(env, { name = "KONG_PG_SSL",
                         value = config.pg_ssl and "ON" or "OFF" })
          tinsert(env, { name = "KONG_PG_SSL_VERIFY",
                         value = config.pg_ssl_verify and "ON" or "OFF" })
        elseif config.database == "cassandra" then
          tinsert(env, { name = "KONG_CASSANDRA_USERNAME",
                         value = config.cassandra_username })
          tinsert(env, { name = "KONG_CASSANDRA_PORT",
                         value = fmt("%d", config.cassandra_port) })
          tinsert(env, { name = "KONG_CASSANDRA_LB_POLICY",
                         value = config.cassandra_lb_policy })
          tinsert(env, { name = "KONG_CASSANDRA_DATA_CENTERS",
                         value = tconcat(config.cassandra_data_centers, ",") })
          tinsert(env, { name = "KONG_CASSANDRA_SSL",
                         value = config.cassandra_ssl and "ON" or "OFF" })
          tinsert(env, { name = "KONG_CASSANDRA_CONSISTENCY",
                         value = config.cassandra_consistency })
          tinsert(env, { name = "KONG_CASSANDRA_REPL_STRATEGY",
                         value = config.cassandra_repl_strategy })
          tinsert(env, { name = "KONG_CASSANDRA_CONTACT_POINTS",
                         value = tconcat(config.cassandra_contact_points, ",") })
          tinsert(env, { name = "KONG_CASSANDRA_SCHEMA_CONSENSUS_TIMEOUT",
                         value = fmt("%d", config.cassandra_schema_consensus_timeout) })
          tinsert(env, { name = "KONG_CASSANDRA_REPL_FACTOR",
                         value = fmt("%d", config.cassandra_repl_factor) })
          tinsert(env, { name = "KONG_CASSANDRA_TIMEOUT",
                         value = fmt("%d", config.cassandra_timeout) })
          tinsert(env, { name = "KONG_CASSANDRA_SSL_VERIFY",
                         value = config.cassandra_ssl_verify and "ON" or "OFF" })
          tinsert(env, { name = "KONG_CASSANDRA_KEYSPACE",
                         value = config.cassandra_keyspace })
        end

        -- Allow plugin config to add extra env vars
        if plugin_config.extra_env then
          for k, v in pairs(plugin_config.extra_env) do
            tinsert(env, { name = k, value = v })
          end
        end

        patches[2] = {
          op = "add",
          path = "/spec/containers/-",
          value = {
            name = "kong-sidecar",
            image = plugin_config.image,
            imagePullPolicy = plugin_config.imagePullPolicy,
            env = env,
            ports = {
             { name = "http", containerPort = 8000, protocol = "TCP" },
             { name = "https", containerPort = 8443, protocol = "TCP" },
             { name = "tcp", containerPort = 7000, protocol = "TCP" },
            },
          },
        }
        if not object.spec.containers then
          -- need to add array member instead of adding *to* it.
          patches[2].path = "/spec/containers"
          patches[2].value = { patches[2].value }
        end

        reply.response.patchType = "JSONPatch"
        reply.response.patch = encode_base64(cjson.encode(patches))
        return { json = reply }
      end,
    },
  },
}
